Template rework

nabobalis opened the pull request:
Trying to fix the templates and tidy them up.

Fixes https://github.com/sunpy/ablog/issues/143
Fixes https://github.com/sunpy/ablog/issues/125 
Fixes https://github.com/sunpy/ablog/issues/108

TODO
- [x] Update release page for 0.11

nabobalis commented on ablog/static/ablog.css:
```diff
@@ -0,0 +1,51 @@
+ul.ablog-archive {
```
TODO: Lint and add to setup so it works.

nabobalis commented on ablog/__init__.py:
```diff
@@ -92,15 +95,13 @@ def setup(app):
 
 def config_inited(app, config):
-    app.config.templates_path.append(get_html_templates_path())
+    app.config.templates_path = [get_html_templates_path()] + app.config.templates_path
```
This should prevent ablog being last on the list?

pradyunsg replied:
This looks... wrong. This puts ablog *before* the user-configured templates, which means even the user can't override these files. These are placed before theme templates too, so themes still won't be able to override the newly moved files.

nabobalis replied:
Ok, so I don't want to add ablog templates to this at all?

pradyunsg replied:
This code has some fairly useful properties and behaviours:

- All the attributes being used here are public, so Sphinx's backwards compatibility policy applies on them, giving us ample time to make changes if Sphinx changes things (deprecation warnings for 2 major versions before removal, which should be more than an year to adapt this code)
- This puts the ablog templates *after* the theme templates, by default, with an escape hatch for users who don't want that. This does depend on them not setting `html_template_bridge` to something super custom -- that's a very small set of users AFAICT and the error message should give such users all the information that they need to act on it.
- This allows themes to *change* where ablog's template injection occurs, by setting a flag in their `theme.conf` file:

  ```
  [ablog]
  inject_templates_after_theme = true
  ```

  This would enable themes like Furo to contain their own `ablog/catalog.html` for
  example.

In the spirit of transparency, I've not tested this code in any way -- and written most of this in a Markdown editor -- so no promises on whether this will work. That said, this will enable basically everything we discussed in #143 and #125 (as long as you move the templates, as has been done already). This should also be a drop-in replacement for users who don't set up custom templates or overrides on top of ablog itself. If they do, they might need to move those paths to match the namespace.

nabobalis replied:
Thanks for all of this, still trying to work out the kinks but your code really helped!

nabobalis commented on ablog/start.py:
```diff
@@ -216,7 +216,7 @@ def w(t, ls=80):
 
 # Add any paths that contain templates here, relative to this directory.
-templates_path = ["{dot}templates", ablog.get_html_templates_path()]
+templates_path = [ablog.get_html_templates_path(), "{dot}templates"]
```
Same here, should be first.

pradyunsg replied:
Users don't need to do this, given that the event handler adds this path already for them.

nabobalis replied:
Right, I will remove this from the setup script.

pradyunsg replied:
Actually, we might want to add a check to reject configurations that do this. If users are doing this, they'll be able to break the way that the suggestion I've made for handling #143 and #125 works in a manner that's gonna be difficult for the user to diagnose (https://github.com/sunpy/ablog/pull/144#discussion_r927425327).

pradyunsg replied:
To elaborate on what I mean, here's the situation elaborated upon:

- Furo adds support for Ablog and declares `inject_templates_after_theme = true` -- to say that the theme's templates should take precedence over Ablog's templates (allowing Furo to have a custom `page.html` and `ablog/catalog.html` for example).
- The above suggested code will make Ablog inject its own templates *after* the theme templates.

If the user does nothing more, everything will work just fine. However, if the user sets this in their `conf.py` file, the Ablog templates will also be added *before* the theme templates (since user templates are injected before theme templates). This would mean that Ablog's own templates would take precedence over the theme's templates. See bullet 1, which expects the exact opposite.

The most annoying part of this for the user is that the build won't fail, since the templates will still likely render correctly -- they'll be getting picked up from the wrong place and cause improper markup to be generated in the final rendered templates!

nabobalis commented on docs/release/ablog-v0.11-released.rst:
```diff
@@ -0,0 +1,23 @@
+
+`Template rework <https://github.com/sunpy/ablog/pull/144>`__.
+
+`Add external links for posts <https://github.com/sunpy/ablog/pull/112>`__ from `Chris Holdgraf <https://github.com/choldgraf>`__.
```
This PR is not merged in but will be when I do the actual release.

pradyunsg commented:
FWIW, one template that you *do not* want to move is `page.html`. That's the template that affects the actual Sphinx templating, since Sphinx renders for (almost) every HTML page and we *want* to override that for alabaster (the default theme).

pradyunsg commented on ablog/__init__.py:
```diff
@@ -62,6 +72,52 @@ def html_page_context(app, pagename, templatename, context, doctree):
 
 
+def builder_inited(app):
+    if not app.config.inject_templates_after_theme:
```
You'll want to register this variable with add_config or whatever that method is called, on the `app` as part of the setup. +

nabobalis replied:
This should now be the case.

pradyunsg commented on ablog/__init__.py:
```diff
@@ -59,6 +72,52 @@ def html_page_context(app, pagename, templatename, context, doctree):
 
 
+def builder_inited(app):
+    if not app.config.inject_templates_after_theme:
+        if not isinstance(app.builder.templates, BuiltinTemplateLoader):
+            raise Exception(
+                "Ablog does not know how to inject templates into with custom "
+                "template bridges. You can use `ablog.get_html_templates_path()` to "
+                "get the path to add in your custom template bridge and set "
+                "`inject_templates_after_theme = False` in your "
+                "`conf.py` file."
+            )
+        if get_html_templates_path() in app.config.templates_path:
+            raise Exception(
+                "Found the path from `ablog.get_html_templates_path()` in the "
+                "`templates_path` variable from `conf.py`. Doing so interferes "
+                "with Ablog's ability to stay compatible with Sphinx themes that "
+                "support it out of the box. Please remove `get_html_templates_path` "
+                "from `templates_path` in your `conf.py` to resolve this."
+            )
```
Hmm... this option seems to be doing the wrong thing.

I'd suggest changing this to `skip_injecting_base_ablog_templates` and changing the conf.py documentation related to this as well (I'll make a separate suggestion for that).

The goal isn't to skip doing any checks when that is set, but to avoid doing *anything* when that is set. This would basically be for users who are using a template bridge that does not work like Sphinx's own template bridge does. That's why we have the `isinstance(..., BuiltinTemplateLoader)` check -- since this logic is highly coupled with the way Sphinx's `BuiltinTemplateLoader` works and we can't know if we'd present esoteric errors with alternatives.

```suggestion
    if app.config.skip_injecting_base_ablog_templates:
        return

    if not isinstance(app.builder.templates, BuiltinTemplateLoader):
        raise Exception(
            "Ablog does not know how to inject templates into with custom "
            "template bridges. You can use `ablog.get_html_templates_path()` to "
            "get the path to add in your custom template bridge and set "
            "`skip_injecting_base_ablog_templates = False` in your "
            "`conf.py` file."
        )
    if get_html_templates_path() in app.config.templates_path:
        raise Exception(
            "Found the path from `ablog.get_html_templates_path()` in the "
            "`templates_path` variable from `conf.py`. Doing so interferes "
            "with Ablog's ability to stay compatible with Sphinx themes that "
            "support it out of the box. Please remove `get_html_templates_path` "
            "from `templates_path` in your `conf.py` to resolve this."
        )
```

pradyunsg replied:
Oh wait, this can't be an early return since this function is also doing `blog_post_pattern` stuff.

Well, the idea still stands, although the specific code suggestion here is wrong. We'd probably want to indent the lines that add the template instead, and only inject templates if the user hasn't explicitly configured a skip.

pradyunsg replied:
It might make sense to make the two of them smaller functions, which would change this function to:

```
def builder_inited(app):
    if not app.config.skip_injecting_base_ablog_templates:
        _inject_base_templates(app)
    if isinstance(app.config.blog_post_pattern, str):
        _inject_matched_blog_posts(app)
```

nabobalis replied:
I moved the `blog_post_pattern ` back to a config-init function where it used to be. This should separate the two parts nicely.

pradyunsg commented on ablog/blog.py:
```diff
@@ -91,45 +91,46 @@ def verify_fn(key, value, config):
-    ("disqus_drafts", False, True),
-    ("disqus_pages", False, True),
-    ("blog_post_pattern", [], True, require_config_type((str, list))),
+    ("inject_templates_after_theme", False, True),
```
```suggestion
    ("skip_injecting_base_ablog_templates", False, True),
```

pradyunsg commented on ablog/start.py:
```diff
@@ -77,6 +77,12 @@ def w(t, ls=80):
 #    'Earth': ('The Blue Planet', 'https://en.wikipedia.org/wiki/Earth),
 # }}
 
+# See https://github.com/sunpy/ablog/pull/144 for the full context
+# This will prevent ablog from overriding any local templates that you might
+# need to use for your project.
+# Default is ``False``.
+# inject_templates_after_theme = False
```
```suggestion
# This will prevent ablog from injecting its own templates into the Sphinx
# build. This is only useful when you have a custom template bridge (rare).
# See https://github.com/sunpy/ablog/pull/144 for the full context.
# skip_injecting_base_ablog_templates = False
```

pradyunsg commented:
So... I did some preliminary testing, starting with `ablog start` from the `main` branch, building with that branch, then switching to this branch and making changes until things build, and finally switching to Furo to see how things build. Notes below.

---

Everything with `main` worked correctly -- ablog start followed by sphinx-build. :)

note 1: The `ablog start`'s conf.py template should get the same treatment as https://github.com/sphinx-doc/sphinx/issues/10056, as well as a cleanup of the stuff it's got that's redundant/useless today (alabaster doesn't need to be imported, os and sys are unused, generated code isn't black-ready etc). We *may* also want to switch to using sphinx-quickstart's templating logic directly, instead of doing `.format` on a hard-coded string in a Python file -- I'll file a PR for this sometime soon.

---

On the first build with the new ablog, users will get the following error:

```
loading pickled environment... failed
failed: No such config value: inject_templates_after_theme

Extension error (ablog):
Handler <function builder_inited at 0x104675c10> for event 'builder-inited' threw an exception (exception: Found the path from `ablog.get_html_templates_path()` in the `templates_path` variable from `conf.py`. Doing so interferes with Ablog's ability to stay compatible with Sphinx themes that support it out of the box. Please remove `get_html_templates_path` from `templates_path` in your `conf.py` to resolve this.)
```

I really like this error message! It provides clear context on what went wrong, why that's a problem and what the user can do. We don't have control over the specific presentation out of Sphinx and that would be my only complaint about this.

However, once I resolved that by following the instructions, you get the following error:

```
Theme error:
An error happened in rendering the page about.
Reason: TemplateNotFound('postcard.html')
```

This is coming out of the `html_sidebars` declaration, and is a very opaque error message. We should improve this.

note 2: We should write a "migrating from {old} ablog to {new} ablog", as a blog post. I'd be happy to write this. :)
note 3: I think it's sensible to add back the "old" files, with the following content:

```py
{{ warning("foo.html is an old template path, that is no longer used by ablog. Please use ablog/foo.html instead.") }}
```

This would mean that users see a warning from Sphinx, that provides guidance, instead of the opaque error above.

---

Changing the theme to Furo hits https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/layout.html. After I added the following to Furo's `theme.conf` (and removed alabaster-specific files from `html_sidebars`), things rendered with Furo's `page.html` template! 🎉 

```ini
[ablog]
inject_templates_after_theme = true
```

This means that it is definitely possible now, to add first-class support for ABlog in Furo and other themes! Right now, quite a few things don't render correctly with Furo and the default templates from ABlog, but that's expected and something for future me to deal with. Right now, I'm *very* excited! :)

nabobalis commented on docs/release/ablog-v0.11-released.rst:
```diff
@@ -0,0 +1,26 @@
+- 1. The template files are now in the `templates/ablog` folder.
+- 2. In theory ablog will now append the tempaltes to the start of the template list, this could break themes without support for it.
+- 3. Minimum version of Python is >=3.8 and Sphinx is >=5.0.
+- 4. Latex building is now not possible anymore.
+     This could change in future but for the first release of 0.11, it is nonfunctional.
```
Double check this is still true.

nabobalis commented:
> note 1: The `ablog start`'s conf.py template should get the same treatment as [sphinx-doc/sphinx#10056](https://github.com/sphinx-doc/sphinx/issues/10056), as well as a cleanup of the stuff it's got that's redundant/useless today (alabaster doesn't need to be imported, os and sys are unused, generated code isn't black-ready etc). We _may_ also want to switch to using sphinx-quickstart's templating logic directly, instead of doing `.format` on a hard-coded string in a Python file -- I'll file a PR for this sometime soon.

Yes that part of `ablog` is pretty old now, it would be great to improve this.
 
> note 2: We should write a "migrating from {old} ablog to {new} ablog", as a blog post. I'd be happy to write this. :) 

That would be incredible!
It is something that would need to exist before a release.
I have created a 0.10 branch and plan add multiple versions to RTD so people can check the old version for a while.

>note 3: I think it's sensible to add back the "old" files, with the following content:
> 
> ```python
> {{ warning("foo.html is an old template path, that is no longer used by ablog. Please use ablog/foo.html instead.") }}
> ```

I will do that in this PR as well but I use the updated html files from this PR.
 
> Changing the theme to Furo hits [pradyunsg/furo@`main`/src/furo/theme/furo/layout.html](https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/layout.html?rgh-link-date=2022-08-05T08%3A53%3A24Z). After I added the following to Furo's `theme.conf` (and removed alabaster-specific files from `html_sidebars`), things rendered with Furo's `page.html` template! 🎉
> 
> ```ini
> [ablog]
> inject_templates_after_theme = true
> ```
> 
> This means that it is definitely possible now, to add first-class support for ABlog in Furo and other themes! Right now, quite a few things don't render correctly with Furo and the default templates from ABlog, but that's expected and something for future me to deal with. Right now, I'm _very_ excited! :)

That is brilliant, I want to thank you for your help again on this PR.

pradyunsg commented on ablog/__init__.py:
```diff
@@ -88,7 +89,7 @@ def config_inited(app):
 
 def builder_inited(app):
-    if app.config.skip_injecting_base_ablog_templates:
+    if isinstance(app.builder, LaTeXBuilder) or app.config.skip_injecting_base_ablog_templates:
```
```
not (isinstance(app.builder, HTMLBuilder) or app.config.skip_injecting_base_ablog_templates))
```

We probably only want blog posts in HTML builds?

nabobalis replied:
Good point.

pradyunsg commented:
@nabobalis Anything I can do to help move this forward?

nabobalis commented:
> @nabobalis Anything I can do to help move this forward?

I am not sure, I forgot about this and need to work out why the CI is broken.

But I think in theory its complete? So maybe just a glance over the Python changes would be ideal.

nabobalis commented on ablog/__init__.py:
```diff
@@ -6,6 +6,9 @@
 from glob import glob
 from pathlib import PurePath
 
+from sphinx.builders.html import StandaloneHTMLBuilder
```
I think this is the correct HTMLBuilder.

nabobalis commented:
Other than the need to fix the wrong setup for the dev build. I think everything is in place.

pradyunsg reviewed (approved):
LGTM, based on a desk skim.

nabobalis commented:
> LGTM, based on a desk skim.

Thanks again. I will branch off main to keep a 0.10 branch around for a bit and then merge this.

choldgraf commented:
🎉

bowentan commented:
Hi, I still got the error for `layout.html` when using the latest `ablog` with furo theme. Is the problem unresolved? Or do I need additional works on configurations?

nabobalis commented:
> Hi, I still got the error for `layout.html` when using the latest `ablog` with furo theme. Is the problem unresolved? Or do I need additional works on configurations?

Hello, this patch never made it to a release version at the moment. 

My plan was to ship this in 0.11 but I got distracted and have not done a release of that.

bowentan commented:
> Hello, this patch never made it to a release version at the moment.
> 
> My plan was to ship this in 0.11 but I got distracted and have not done a release of that.

Oh i see.. Thanks!

Jaharmi commented:
I am also seeing this on a new site / test setup I'm working on. It looks like there is no 0.11 release yet.

nabobalis commented:
Yes unfortunately that is the case. Real life has got in my way recently and that will not change for me until Christmas.

nabobalis commented:
Sorry it has taken me this long but I released ablog 0.11.0rc1.

I did not want to release such a large change without at least a brief RC period.

So if those who wanted to try this out for me, that would be great. 

I plan to test some of the projects I maintain that use ablog as well but its a very limited test.

choldgraf commented:
Nice - sounds good I will give it a shot and open issues
